### PR TITLE
[MHLO] Add [un]squeeze op patterns

### PR DESCRIPTION
**Pull request by fortianyou**
See RFC https://github.com/llvm/torch-mlir/issues/999

Co-authored-by: Bairen Yi [yibairen.byron@bytedance.com](mailto:yibairen.byron@bytedance.com)
Co-authored-by: Jiawei Wu [xremold@gmail.com](mailto:xremold@gmail.com)
Co-authored-by: Tianyou Guo [tianyou.gty@alibaba-inc.com](mailto:tianyou.gty@alibaba-inc.com)
Co-authored-by: Xu Yan [yancey.yx@alibaba-inc.com](mailto:yancey.yx@alibaba-inc.com)
Co-authored-by: Ziheng Jiang [ziheng.jiang@bytedance.com](mailto:ziheng.jiang@bytedance.com)

### CODE REVIEW

**Comment by fortianyou** [comment]
Follow up https://github.com/llvm/torch-mlir/pull/1090

Ready for review @silvasean @powderluv @Vremold @ZihengJiang @byronyi @Yancey1989

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 1]
```diff
@@ -49,17 +59,33 @@ SmallVector<Value, 4> getDimSizesOfTensor(PatternRewriter &rewriter,
...
         rewriter.create<tensor::DimOp>(loc, value, d)));
   }
   return dimSizes;
 }
 
+SmallVector<Value, 4> getDimSizesOfTensor(PatternRewriter &rewriter,
+                                          Operation *op, Value value) {
+  auto valueTy = value.getType().dyn_cast<RankedTensorType>();
+  if (!valueTy) {
+    op->emitOpError("getDimSizesOfTensor(): the input is not a ranked tensor");
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:78]: Either rewriter.notifyMatchFailure if this is a limitation of the pattern or the check should be in the verifier of the op.

Could the return type be FailureOr<...> ?

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 2]
```diff
@@ -316,6 +342,150 @@ bool ConvertAtenViewOp<AtenReshapeOp>::getAtenViewOpSizes(
...
+  // tensor) are specified with unsqzDims. Indices must be in-order, and in
+  // range of tensor rank. Thus, unsqueeze a rank 1 tensor with {0, 2}, {0, 1,
+  // 3}, {0, 1, 2} are all valid dimension sets, but {0, 3}, {2} are not.
+  auto dimSizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto rank = dimSizes.size();
+  size_t newRank = rank + inputUnsqzDims.size();
+  auto unsqzDims = toPositiveDims(inputUnsqzDims, newRank);
+  for (size_t k = 0; k < unsqzDims.size(); ++k) {
+    if (k > 1 && unsqzDims[k] <= unsqzDims[k - 1]) {
+      op->emitOpError("Unsqueeze dimensions must be specified in order.");
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:361]: Same question wrt matchFailure vs emitOpError here and elsewhere.

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 3]
```diff
@@ -316,6 +342,150 @@ bool ConvertAtenViewOp<AtenReshapeOp>::getAtenViewOpSizes(
...
+  //
+  // The position indices (must be high to low dimension number of the returned
+  // tensor) are specified with unsqzDims. Indices must be in-order, and in
+  // range of tensor rank. Thus, unsqueeze a rank 1 tensor with {0, 2}, {0, 1,
+  // 3}, {0, 1, 2} are all valid dimension sets, but {0, 3}, {2} are not.
+  auto dimSizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto rank = dimSizes.size();
+  size_t newRank = rank + inputUnsqzDims.size();
+  auto unsqzDims = toPositiveDims(inputUnsqzDims, newRank);
+  for (size_t k = 0; k < unsqzDims.size(); ++k) {
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:359]: Unless unsqzDims change during loop iterations, don't check in comparison (https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop)

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 4]
```diff
@@ -316,6 +342,150 @@ bool ConvertAtenViewOp<AtenReshapeOp>::getAtenViewOpSizes(
...
+}
+
+template <>
+LogicalResult ConvertAtenOp<AtenSqueezeOp>::matchAndRewrite(
+    AtenSqueezeOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto self = adaptor.self();
+  auto selfTy = self.getType().template cast<RankedTensorType>();
+  if (!selfTy)
+    return op.emitError("Only ranked tensor types supported in MHLO");
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:402]: Note: for errors they should follow https://llvm.org/docs/CodingStandards.html#error-and-warning-messages (no capital at start, sentence fragment)

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 5]
```diff
@@ -316,6 +342,150 @@ bool ConvertAtenViewOp<AtenReshapeOp>::getAtenViewOpSizes(
...
+}
+
+template <>
+LogicalResult ConvertAtenOp<AtenSqueezeDimOp>::matchAndRewrite(
+    AtenSqueezeDimOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto self = adaptor.self();
+  auto selfTy = self.getType().template cast<RankedTensorType>();
+  if (!selfTy)
+    return op.emitError("Only ranked tensor types supported in MHLO");
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:438]: That's probably true of lowering, but not of the MHLO opset itself.

File: lib/Conversion/TorchToMhlo/ViewLikeOps.cpp [hunk 6]
```diff
@@ -316,6 +342,150 @@ bool ConvertAtenViewOp<AtenReshapeOp>::getAtenViewOpSizes(
...
+    return rewriter.notifyMatchFailure(
+        op, "The rank of tensor must be greater than 0");
+  }
+
+  dim = toPositiveDim(dim, rank);
+  if (selfTy.getShape()[dim] != 1) {
+    if (selfTy.getShape()[dim] == ShapedType::kDynamicSize) {
+      return rewriter.notifyMatchFailure(
+          op, "The size of the dimension being squeezed is can't be unknown");
+    } else {
```
Comment (jpienaar) [inline lib/Conversion/TorchToMhlo/ViewLikeOps.cpp:455]: You can drop if post if that returns.

**Review by jpienaar — COMMENTED** [review]
Nice, thanks

**Comment by fortianyou** [comment]
@jpienaar Thanks for advising, I have updated the PR to conform to LLVM coding standard.

**Comment by jpienaar** [comment]
> @jpienaar Thanks for advising, I have updated the PR to conform to LLVM coding standard.

Thanks!

**Pull request merged**